packages/material-ui-lab/src/TabContext/TabContext.js

```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+
+const Context = React.createContext(null);
+if (process.env.NODE_ENV !== 'production') {
+  Context.displayName = 'TabContext';
+}
+
+function useUniquePrefix() {
```

Comment:
Minor question: did you redefine this instead of importing `unstable_useId` because it's "unstable"? I think it'd work the same as this if you just call `unstable_useId()` or `unstable_useId(null)`

Reply:
I redefined it because the semantics are different (with useOpaqueIdentifier in mind). `useId` creates a value that should be passed to ´id` attribute (or an IDREF attribute for that matter). `useUniquePrefix` creates a string that is unique and can be used to append another string.

So, yes, they do work the same way right now but that might change in the future.

Reply:
I'm not sure to follow the difference between `unstable_useId` and `useUniquePrefix`. For instance, what's the difference with the usage in the autocomplete, it requires the following properties

- uniqueness, we apply it as an id:
https://github.com/mui-org/material-ui/blob/6e30e5ccdaf256ff33cc27ede9e967feb0a8e7b4/packages/material-ui-lab/src/useAutocomplete/useAutocomplete.js#L921
- to support prefixing:
https://github.com/mui-org/material-ui/blob/6e30e5ccdaf256ff33cc27ede9e967feb0a8e7b4/packages/material-ui-lab/src/useAutocomplete/useAutocomplete.js#L908

@NMinhNguyen Long term, should we use `useOpaqueIdentifier` inside `useId`?

Reply:
> @NMinhNguyen Long term, should we use `useOpaqueIdentifier` inside `useId`?

That's my understanding, yeah. Or maybe even just using it directly?

Reply:
> I'm not sure to follow the difference between unstable_useId and useUniquePrefix.

useId is something you pass directly to an `id` attribute. No other valid usage is implied. `usePreifx` returns a value that can be concatenated at any point in time.

If you currently use `useId` to concatenate strings than this blocks us from replacing `useId` with a different implementation.

Reply:
There is one element I don't understand. Why does it matter that the randomly generated string is used as an id, a prefix of id, or any other DOM attribute? Maybe we should have our own `useOpaqueIdentifier`? A hook we would replace with the one React is working on once our peer dependency target allows it?

Reply:
> Why does it matter that the randomly generated string is used as an id, a prefix of id, or any other DOM attribute?

You assume one is a `string` (useUniquePrefix) while you don't assume anything about the value you directly pass to the `id` (useId) attribute. Could be a number, a function, an object etc. All you care about is that react accepts this as an id attribute.

Reply:
I dig a bit into the React's code base. The returned value of `useOpaqueIdentifier()` is: 

```jsx
export opaque type OpaqueIDType =
  | string
  | {
      toString: () => string | void,
      valueOf: () => string | void,
    };
```
https://github.com/facebook/react/blob/aa88589d0b56521512e864243744513ee1f93a81/packages/react-test-renderer/src/ReactTestHostConfig.js#L48-L53.

What if we set a new constraint, the input and output of our unique identifier should be a string?

Reply:
Check out https://en.wikipedia.org/wiki/Opaque_data_type. Maybe this clears things up for you. Going to mark this as resolved since we can fix current usage of useId in another PR. useUniqueString and useId have different semantics.

Reply:
I don't understand what we are talking about. This is too abstract for me to comprehend.

Reply:
I don't see why we would want to accept a number, object, or function. As far as I'm aware of, useId only accept a string as a first argument. Anything else is cast to a string and would raise a TypeScript or PropType warning.

Reply:
@oliviertassinari I think what @eps1lon is saying that React's `useOpaqueIdentifier` returns a special data type that is understood by React's DOM elements like `div` etc, and that it's not necessarily a string. Which means that it can't be easily concatenated with another string. But ultimately it gets converted to a string when it's actually committed to the DOM. Whereas the semantics of this new hook `useUniquePrefix` is that it only returns a string and is thus safe to use in concatenation. The fact that the two hooks currently return a string is more of a coincidence/implementation detail.

Reply:
@NMinhNguyen Thanks for the explanation, it makes sense. If I understand correctly we have two cases:

- a prefix string (useUniquePrefix). Needed for TabPanel (**new**), Autocomplete, Rating.
- a full id (useId/useOpaqueIdentifier). Needed for Tooltip, Autocomplete, Rating, Radio.

Would it make sense then to make `useUniquePrefix()` a shared util?

Reply:
> As far as I'm aware of, useId only accept a string as a first argument. Anything else is cast to a string and would raise a TypeScript or PropType warning.

Yes, any released version of react casts the value in `id={value}` to a string. But not the one that implements `useOpaqueIdentifier`. In a world where this implementation doesn't exist, this discussion is fruitless. But we might as well anticipate that change. It doesn't cost use anything other than having a discussion about duplicating a few bytes.

Reply:
We might want to raise to the React team about the useOpaqueIdentifier prefix use case.  If we take an autocomplete with 100,000 virtualized options, we couldn't create that many hooks.

Reply:
Got you covered: https://github.com/reactjs/rfcs/pull/32#discussion_r419501533

Reply:
Though I didn't look at the autocomplete case. It might be even more compelling.

Reply:
Awesome, for the autocomplete, I think that the concern is that it can often have x10 more items than a tab. The ids are required for https://www.w3.org/TR/wai-aria-practices/#kbd_focus_activedescendant.

Reply:
@eps1lon given Dan's https://github.com/reactjs/rfcs/pull/32#issuecomment-610677158  , do you think it's worthwhile moving that convo to the main React repo instead? Because it relates to the current implementation.